text/0000-structural-records.md

```diff
@@ -0,0 +1,1262 @@
+  
+  ```rust
+  {
+      alpha: 42,
+      beta: true,
+      gamma: "Dancing Ferris"
+  }
+  ```
+
++ For `Hash`, the same ordering of the fields as before is used and then
+  `self.the_field.hash(state)` is called on each field in that order.
+  For example:
+
+  ```rust
+  self.alpha.hash(state);
+  self.beta.hash(state);
+  self.gamma.hash(state);
+  ```
+
+For auto traits (e.g. `Send`, `Sync`, `Unpin`), if all field types implement
+the auto trait, then the structural record does as well.
+For example, if `A` and `B` are `Send`, then so is `{ x: A, y: B }`.
+
+A structural record is `Sized` if all field types are.
+If the lexicographically last field of a structural record is `!Sized`,
+then so is the structural record. If any other field but the last is
+`!Sized`, then the type of the structural record is not well-formed.
+
+### Implementations and orphans
+
+[RFC 2451]: https://github.com/rust-lang/rfcs/pull/2451
+
+It is possible to define your own implementations for a structural record.
+The orphan rules that apply here are those of [RFC 2451] by viewing a structural
+record as a positional tuple after sorting the elements lexicographically.
+
+For example, if a `trait` is crate-local, we may implement it for a record:
+
+```rust
+trait Foo {}
+
+impl Foo for { alpha: bool, beta: u8 } {}
+```
+
+Under 2451, we can also write:
+
+```rust
+struct Local<T>(T);
+
+impl From<()> for Local<{ alpha: bool, beta: u8 }> { ... }
+```
+
+This is valid because `Local` is considered a local type.
+
+However, a structural record itself isn't a local type, so you cannot write:
+
+```rust
+struct A;
+
+impl From<()> for { alpha: A, beta: u8 } { ... }
+```
+
+This is the case even though `A` is a type local to the crate.
+
+The behaviour for inherent implementations is also akin to tuples.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Grammar
+[grammar]: #grammar
+
+Given:
+
+```rust
+field ::= ident | LIT_INTEGER ;
+```
+
+We extend the expression grammar with:
+
+```rust
+expr ::= ... | expr_srec ;
+
+expr_srec ::= "{" field_init "," field_inits_tail? "}" ;
+field_inits_tail ::= field_inits ","? | (field_inits ",")? DOTDOT expr ;
+field_inits ::= field_init ("," field_init)* ;
+field_init ::= ident | field ":" expr ;
+```
+
+Note that this grammar permits `{ 0: x, 1: y }`.
+We do so to increase reuse in parsing, to improve consistency,
+and to simplify for macro users.
+This is also applied in the grammars below.
+
+We extend the pattern grammar with:
+
+```rust
+pat ::= ... | pat_srec ;
+
+pat_srec ::= "{" pat_srec_start "}" ;
+pat_srec_start ::= DOTDOT | pat_field "," pat_srec_tail
+pat_srec_tail ::= pat_fields ","? | (pat_fields ",")? DOTDOT? ;
+pat_fields ::= pat_field ("," pat_field)* ;
+pat_field ::= BOX? binding_mode? ident | field ":" pat ;
+```
+
+Finally, we extend the type grammar with:
+
+```rust
+ty ::= ... | ty_srec ;
+
+ty_srec ::= "{" (ty_field ",")+ (ty_field ","?)? "}" ;
```

Comment:
I don't think types can start with `{` syntactically.
There are multiple already existing ambiguities where parser assumes that types cannot start with `{`, and upcoming const generics introduce one more big ambiguity.
https://github.com/rust-lang/rfcs/pull/2102 uses `struct { field: Type, ... }` to solve this.

Reply:
I'm not sure about expressions/patterns.
Unlimited lookahead may also be required, need to check more carefully.

Reply:
> There are multiple already existing ambiguities where parser assumes that types cannot start with `{`

Such as? (examples please...)

Const generics allow literals and variables as expressions but anything else needs to be in `{ ... }`.
This is not ambiguous with structural records (because the one-field-record requires a comma...) but requires lookahead.

>  Unlimited lookahead may also be required, need to check more carefully.

Scroll down ;) It's discussed in the sub-section "Backtracking".

Reply:
>It's discussed in the sub-section "Backtracking".

It's a huge drawback.

Reply:
>Such as? (examples please...)

Nothing that can't be solved by infinite lookahead, but here's "how to determine where `where` clause ends":
```
fn f() where PREDICATE1, PREDICATE2, { a: ...
```
Does `{ a: ...` belong to the function body or to the third predicate?

Reply:
> It's a huge drawback.

I agree that it is a drawback (as noted in the text).
However, I don't, even nearly, agree that the drawback is *huge*.

As @eddyb noted in https://github.com/rust-lang/rfcs/pull/2544#issuecomment-422250552, LL/LR have overstayed their welcome and people can handle this with GLL / easily with recursive decent / also easily with parser combinators.

> Does `{ a: ...` belong to the function body or to the third predicate?

None of these are particularly likely. If it belongs to the function body, then it is a type ascription of an identifier; that seems pathological to me. Thus, I would tune the parser towards first parsing the latter, but this can be experimented with of course. I'd also note that you can determine after the first field which it is... That said, I've included this in the relevant section.

Reply:
"Not particularly likely" doesn't mean that the parser doesn't need to handle it, though. Even if "only" for better, i.e. not completely misleading, error messages, it probably realistically needs to do something with this sort of ambiguity anyway.

Reply:
@H2CO3 oh sure; the parser must handle it, but the perils of backtracking is that it would considerably slow down parsing or produce bad error messages. What I'm saying is that the code paths in the parser that would handle this are pathological, so the slow downs are unlikely.

There's also no *ambiguity* here at all, but lookahead / backtracking / GLL may be required.

If we ever use the [GLL crate](https://github.com/lykenware/gll/) in the compiler then we don't even need backtracking... it provides a correct algorithm to parse context-free-grammars and can produce good error messages for these sort of things because it produces a parse forest that you can catch certain patterns on.

Reply:
The fact that advanced formalisms exist doesn't necessarily mean that we should use them.
I many respects the simpler means the better.
Officially abandoning `LL(n)` is a separate decision that shouldn't be done in comments to a mostly unrelated RFC.

(That said, for this RFC disambiguation can also be done using cover grammar `TYPE_OR_EXPR`.
We parse `{ IDENT: TYPE_OR_EXPR` and look at the next token, if it's `,` then we have a struct record type.)

(Nit: GLL *is* a backtracking of sorts, just with fancy memoization making is `O(n^3)` rather than exponential).

Reply:
> The fact that advanced formalisms exist doesn't necessarily mean that we should use them.
> I many respects the simpler means the better.

Well there are advantages to both, and we should decide what trade-off is best for Rust.

> Officially abandoning `LL(n)` is a separate decision that shouldn't be done in comments to a mostly unrelated RFC.

Abandoning `LL(k)` must be made for a reason, and that reason would be some feature or change we would like to make that forces us to do it. I don't think (correct me if that is inaccurate) we ever stated explicitly that `LL(k)` is a design goal so if so it doesn't need explicit and official abandonment either.

So I think either we abandon `LL(k)` on this RFC or in the https://github.com/rust-lang/rfcs/pull/2544 RFC, or we accept neither and keep `LL(k)`.

> (That said, for this RFC disambiguation can also be done using cover grammar `TYPE_OR_EXPR`.
> We parse `{ IDENT: TYPE_OR_EXPR` and look at the next token, if it's `,` then we have a struct record type.)

Yeah, that's the idea I had also.

> (Nit: GLL _is_ a backtracking of sorts, just with fancy memoization making is `O(n^3)` rather than exponential).

Fair enough ;)

Reply:
Even if you disregard the fact that this syntax would necessitate backtracking, it is confusing for readers IMO. Why not require a `struct` keyword or something?

Reply:
If it's going to be drawbacks of this pattern for the structural records, either for slowing parser or cause types can't start with `{` syntactically, couldn't we use `.{` instead?

Inspiration by Zig programming language

Reply:
Actually, I've been thinking about this recently and I've come to the conclusion that if we lean into the ambiguity, then this is a non-issue.

The need for backtracking here stems from ambiguity with type ascription. This can be fixed by simply assuming that `{ ident:` is always going to be a structural record, either in expression or type position (also solves the const generics issue). That is, if given `{ foo: Vec<u8> }` in an expression context, we will attempt parsing this as a structural record, and then fail at `<` with a parse error (backtracking can be done inside diagnostics only if we want to). This way, we have retained `LL(k)` (standing at `{`, the `:` is 2 tokens away and so therefore `k = 2` in this instance).

If a user wants to disambiguate in favor of ascription, they can write `{ (ident: Type) }`. Disambiguating in this way should not be a problem as `{ ident: Type }` is not a very useful thing to write. Specifically, in the case of `PREDICATE2, { a: ...`, this should not be a problem because the return type itself is a coercion site and informs type inference.

Reply:
Why would it fail at `<`?  If you're trying to parse `Vec<u8>` as an expression, wouldn't the `<` be parsed as less-than, and you only fail when you see `>` (and in general arbitrarily far away)?

It seems like the same issue as when we tried to get rid of turbofish...

Reply:
@comex Oh; yeah, you're right. It should fail at `>`. Still, we can take a snapshot before trying to parse the expression after `ident: ` and backtrack if it turned out to be a type and provide good `MachineApplicable` diagnostics here.

> It seems like the same issue as when we tried to get rid of turbofish...

Well in the sense that turbofish would also be unnecessary if we are OK with backtracking, but it seems like some people object to backtracking, so I've come up with a solution using `LL(k)` at the cost of not being able to recognize `{ binding: Type }` as a block with a tail-expression ascribing `binding`. Because such code is, in my view, pathological, that should be fine. It's much less of an issue than turbofish from a learnability POV imo.